libpod/image/image.go

```diff
@@ -1076,3 +1078,144 @@ func (i *Image) Comment(ctx context.Context, manifestType string) (string, error
+	layer := i.TopLayer()
+	dependencyList := returnDependencyTree(layerMap, repoTag, layer)
+
+	var buffer bytes.Buffer
+	printTree(&buffer, dependencyList)
+
+	return &buffer, err
+}
+
+func printTree(buffer *bytes.Buffer, treeStack []TreeImage) {
+
+	prefixSpace := 0
+	for i := len(treeStack) - 1; i >= 0; i-- {
+		for p := 0; p < prefixSpace; p++ {
+			buffer.WriteString(fmt.Sprintf(" "))
+		}
+		buffer.WriteString(fmt.Sprintf("└─ ID: %s  Tags: %s VirtualSize: %v\n",
+			treeStack[i].ID[:12],
+			treeStack[i].RepoTags,
+			units.HumanSizeWithPrecision(float64(treeStack[i].Size), 4)),
+		)
+
+		prefixSpace++
+	}
+}
+
+func (i *Image) layerToImageMap(ctx context.Context) (map[string]*TreeImage, error) {
+	layerMap := make(map[string]*TreeImage)
+	layers, _ := i.imageruntime.store.Layers()
+	for _, layer := range layers {
+		existingLayer, ok := layerMap[layer.ID]
+		if !ok {
+			layerMap[layer.ID] = &TreeImage{
+				ID:       layer.ID,
+				Size:     layer.CompressedSize,
+				ParentID: layer.Parent,
+			}
+		} else {
+			return nil, fmt.Errorf("duplicate %s layer found in image store", existingLayer.ID)
+		}
+	}
+	return layerMap, nil
+}
+
+func (i *Image) addImageToLayerInfo(ctx context.Context, layerMap map[string]*TreeImage) (map[string]*TreeImage, error) {
+	images, err := i.imageruntime.GetImages()
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting images from store")
+	}
+
+	for _, image := range images {
+
+		layerid := image.TopLayer()
+		imgLayer, err := image.imageruntime.store.Layer(layerid)
+		if err != nil {
+			return nil, err
+		}
+
+		for imgLayer != nil {
+			existingLayer, ok := layerMap[imgLayer.ID]
+			if !ok {
+				return nil, fmt.Errorf("layer %s not found ", imgLayer.ID)
+			} else {
+				names := image.Names()
+				existingLayer.RepoTags = append(existingLayer.RepoTags, names...)
+			}
+			imgLayer = image.getLayer(imgLayer.Parent)
+		}
+	}
+	return layerMap, nil
+}
+
+func (i *Image) getLayer(layerID string) *storage.Layer {
+	layers, err := i.imageruntime.store.Layers()
+	if err != nil {
+		return nil
+	}
+
+	for _, layer := range layers {
+		if layerID == layer.ID {
+			return &layer
+		}
+	}
+	return nil
+}
+
+func returnDependencyTree(layerMap map[string]*TreeImage, repoTag string, layerID string) []TreeImage {
+	var treeStack []TreeImage
+	previous := TreeImage{}
+	layerInfo, ok := layerMap[layerID]
+	for ok {
+		appended := false
+		// First loop
+		if len(previous.RepoTags) == 0 {
+			previous.ID = layerInfo.ID
+			previous.RepoTags = []string{repoTag}
+		}
+
+		if previous.RepoTags[0] == layerInfo.RepoTags[0] {
+			previous.Size = previous.Size + layerInfo.Size
+		} else if layerInfo.Size != 0 {
+			treeStack = append(treeStack, previous)
+			previous = TreeImage{}
+			previous.ID = layerInfo.ID
+			previous.RepoTags = []string{layerInfo.RepoTags[0]}
+
+		}
+
+		ok = false
+		if layerInfo.ParentID != "" {
+			layerInfo, ok = layerMap[layerInfo.ParentID]
```

Comment:
here, by `ParentID == " " ` means we reached root layer and we need to break the loop. Since the map is already built, there is no chance of race condition here.